kube-runtime/src/observer.rs

```diff
@@ -0,0 +1,114 @@
+use backoff::ExponentialBackoff;
+use futures::{future::ready, Stream, StreamExt};
+use kube_client::{
+    api::{Api, ListParams},
+    core::Resource,
+};
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+use crate::{
+    utils,
+    watcher::{watcher, Error, Event, Result},
+};
+
+
+/// A simple observer around a watcher with error handling and retry backoff
+///
+/// # Error handling
+///
+/// An `Observer` sets a sensible default backoff policy for all watch events and will retry
+/// (with expotential backoff) from transient `watcher` failures until the retry policy is breached.
+///
+/// To configure the retry policy use `Observer::backoff`.
+///
+/// Note that while all errors will be retried, some represent the need for user action to recover:
+///
+/// - 404 `ErrorResponse`(watching invalid / missing api kind/group for `K`)
+/// - 403 `ErrorResponse` (missing list + watch rbac verbs for `K`)
+pub struct Observer<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+{
+    // temporary builder params
+    api: Api<K>,
+    listparams: Option<ListParams>,
```

Comment:
This could just be a `ListParams` that we init to default in `new`

Reply:
the goal is to hide listparams in the builder unless users explicitly need to customise it, so it's defaulted later if it's not passed.